dev/src/bulk-writer.ts

```diff
@@ -618,18 +684,55 @@ export class BulkWriter {
    *   console.log('Executed all writes');
    * });
    */
-  async flush(): Promise<void> {
+  flush(): Promise<void> {
     this.verifyNotClosed();
-    const trackedBatches = this.batchQueue;
-    const writePromises = trackedBatches.map(batch => batch.awaitBulkCommit());
-    this.sendReadyBatches();
-    await Promise.all(writePromises);
+
+    // Copy the pending ops at the time flush() was called.
+    const pendingOps: Array<Promise<void>> = [];
+    this._pendingOps.forEach(op => pendingOps.push(op));
+
+    return this._flush(pendingOps);
+  }
+
+  private async _flush(pendingOps: Array<Promise<void>>): Promise<void> {
+    const existingBatches = this._batchQueue;
+    existingBatches.map(batch => batch.markReadyToSend());
+    let retriesCompleted = false;
+    let isFirstLoop = true;
+
+    // Send all scheduled operations on the BatchQueue first. Afterwards,
+    // send all retries that the initial batches generated in a loop. This way,
+    // operations scheduled after flush() will not be sent until the retries
+    // are completed.
+    while (!retriesCompleted) {
+      const before = this._pendingBatches.size;
+      if (isFirstLoop) {
+        this.sendReadyBatches();
+        isFirstLoop = false;
+      } else {
+        this._retryBatchQueue.forEach(batch => batch.markReadyToSend());
+        this.sendReadyBatches(this._retryBatchQueue);
+      }
+      const after = this._pendingBatches.size;
```

Comment:
private async _flush(pendingOps: Array<Promise<void>>): Promise<void> {
```suggestion
  private async _flush(pendingOps: Array<Promise<void>>): Promise<void> {
    let batchQueue = this._batchQueue;

    // Send all scheduled operations on the BatchQueue first. Afterwards,
    // send all retries that the initial batches generated in a loop. This way,
    // operations scheduled after flush() will not be sent until the retries
    // are completed.
    while (batchQueue.length > 0) {
      batchQueue.forEach(batch => batch.markReadyToSend());
      this.sendReadyBatches(batchQueue);
      await Promise.all(this._pendingBatches);
      batchQueue = this._retryBatchQueue;
    }

    // Make sure user promises resolve before flush() resolves.
    return voidPromiseNoThrow(Promise.all(pendingOps));
  }
```

How about this? (untested)